src/context.rs

```diff
@@ -0,0 +1,623 @@
+    ///     2. Select all [`ActionValue`]s with the most significant [`ActionState`] and combine them using the
+    ///        [`ActionSettings::accumulation`] strategy.
+    ///     3. Convert the combined value to [`ActionOutput::DIM`] using [`ActionValue::convert`].
+    ///     4. Apply all action-level [`InputModifier`]s.
+    ///     5. Evaluate all action-level [`InputCondition`]s, combining their results based on their [`InputCondition::kind`].
+    ///     6. Convert the final value to [`ActionOutput::DIM`] again using [`ActionValue::convert`].
+    ///     7. Apply the resulting [`ActionState`] and [`ActionValue`] to the action entity.
+    ///     8. If the final state is not [`ActionState::None`], consume the binding input value.
+    ///
+    /// This logic may look complicated, but you don't have to memorize it. It behaves surprisingly intuitively.
+    fn add_input_context<C: Component>(&mut self) -> &mut Self {
+        self.add_input_context_to::<PreUpdate, C>()
+    }
+
+    /// Like [`Self::add_input_context`], but allows specifying the schedule
+    /// in which the context's actions will be evaluated.
+    ///
+    /// For example, if your game logic runs inside [`FixedMain`](bevy::app::FixedMain), you can set the schedule
+    /// to [`FixedPreUpdate`]. This way, if the schedule runs multiple times per frame, events like [`Started`] or
+    /// [`Completed`] will be triggered only once per schedule run.
+    fn add_input_context_to<S: ScheduleLabel + Default, C: Component>(&mut self) -> &mut Self;
+}
+
+impl InputContextAppExt for App {
+    fn add_input_context_to<S: ScheduleLabel + Default, C: Component>(&mut self) -> &mut Self {
+        debug!(
+            "registering `{}` for `{}`",
+            any::type_name::<C>(),
+            any::type_name::<S>(),
+        );
+
+        let id = self.world_mut().register_component::<Actions<C>>();
+        let mut registry = self.world_mut().resource_mut::<ContextRegistry>();
+        if let Some(contexts) = registry
+            .iter_mut()
+            .find(|c| c.schedule_id == TypeId::of::<S>())
+        {
+            debug_assert!(
+                !contexts.actions_ids.contains(&id),
+                "context `{}` shouldn't be added more then once",
+                any::type_name::<C>()
+            );
+            contexts.actions_ids.push(id);
+        } else {
+            let mut contexts = ScheduleContexts::new::<S>();
+            contexts.actions_ids.push(id);
+            registry.push(contexts);
+        }
+
+        self.register_required_components::<C, ContextPriority<C>>()
+            .add_observer(register_instance::<C, S>)
+            .add_observer(remove_context::<C, S>);
+
+        self
+    }
+}
+
+/// Tracks registered input contexts for each schedule.
+///
+/// In Bevy, it’s impossible to know which schedule is used inside a system,
```

Comment:
I bet we could expose this in the future.

Reply:
Ah, it would be great! :slightly_smiling_face: